Add multus-sriov documentation (#927)
rancher/rke2: docs/install/network_options.md

```

For more information about the full options of the Cilium config please refer to the [rke2-charts](https://github.com/rancher/rke2-charts/blob/main-source/packages/rke2-cilium/charts/values.yaml).

# Using Multus

> Starting with v1.21 it is possible to deploy Multus-cni. Note that this is for advanced users.

Reviewer: ```suggestion
Starting with RKE2 v1.21 it is possible to deploy the Multus CNI meta-plugin. Note that this is for advanced users.
```

Multus can not be deployed standalone. It always requires at least one convential CNI plugin that fulfill the Kubernetes cluster network requirements such as canal, calico or cilium. That CNI plugin becomes the default for Multus and will be used for all the pods. Therefore, to deploy multus in rke2, the user must fill the `--cni` flag with `multus, $CNIPlugin`, where $CNIPlugin can be either canal, calico or cilium. Note that multus must always be in the first position of the list.

For more information about Multus refer to [multus-cni](https://github.com/k8snetworkplumbingwg/multus-cni/tree/master/docs)

## Using Multus with the containernetworking plugins

Any CNI plugin could be used as secondary CNI plugin for Multus. These will manage the possible additional network interfaces attached per pod. It is common to use the CNI plugins, maintained by the containernetworking team, as secondary CNI plugins for Multus (e.g. macvlan CNI). Therefore, these plugins get automatically deployed when installing multus. For more information about these plugins, check the [containernetworking plugins](https://www.cni.dev/plugins/current).

> To use any of these plugins, a proper NetworkAttachmentDefinition object will need to be created, defining the new available network. Subsequently, in the pod annotations, we can point to that new network, so that Multus will provide an extra interface to that pod. Please use this example as reference to macvlan cni plugin with multus [example](https://github.com/k8snetworkplumbingwg/multus-cni/blob/master/docs/quickstart.md#storing-a-configuration-as-a-custom-resource)

Reviewer: ```suggestion
To use any of these plugins, a proper NetworkAttachmentDefinition object will need to be created to define the configuration of the secondary network. The definition is then referenced by pod annotations, which Multus will use to provide extra interfaces to that pod. An example using the macvlan cni plugin with Multus is available [in the multus-cni repo](https://github.com/k8snetworkplumbingwg/multus-cni/blob/master/docs/quickstart.md#storing-a-configuration-as-a-custom-resource).
```

> ## Using Multus with SRIOV

Reviewer: ```suggestion
## Using Multus with SR-IOV
```

> Using the SRIOV-cni as part of Multus, can help with dataplane acceleration use cases, providing an extra interface in the pod that can achieve very high throughput. SRIOV will not work in all environments and there are several requirements that must be fulfilled to consider the node as sriov capable:

Reviewer: ```suggestion
Using the SR-IOV CNI with Multus can help with data-plane acceleration use cases, providing an extra interface in the pod that can achieve very high throughput. SR-IOV will not work in all environments and there are several requirements that must be fulfilled to consider the node as SR-IOV capable:
```

> * Physical NIC must support SRIOV (e.g. by checking /sys/class/net/$NIC/device/sriov_totalvfs)

Reviewer: ```suggestion
* Physical NIC must support SR-IOV (e.g. by checking /sys/class/net/$NIC/device/sriov_totalvfs)
```

* The host operating system must activate IOMMU virtualization
* The host operating system includes drivers capable of doing sriov (e.g. i40e, vfio-pci, etc)

> As the sriov-cni will become a secondary CNI plugin for multus, it is mandatory to have selected multus as the cni. Then, to install the SRIOV components, we must turn to the Helm Charts of Rancher, where the SRIOV charts are available. For more information [rancher Helm Charts](https://rancher.com/docs/rancher/v2.x/en/helm-charts/)

Reviewer: ```suggestion
The SR-IOV CNI plugin cannot be used as the default CNI plugin for Multus; it must be deployed alongside both Multus and a traditional CNI plugin. The SR-IOV CNI helm chart can be found in the `rancher-charts` Helm repo. For more information see [Rancher Helm Charts documentation](https://rancher.com/docs/rancher/v2.x/en/helm-charts/).
```

> After installing the sriov charts, the sriov operator will be deployed. Then, the user must specify what nodes in the cluster are sriov capable by providing the label feature.node.kubernetes.io/network-sriov.capable=true:

Reviewer: ```suggestion
After installing the SR-IOV CNI chart, the SR-IOV operator will be deployed. Then, the user must specify what nodes in the cluster are SR-IOV capable by labeling them with `feature.node.kubernetes.io/network-sriov.capable=true`:
```

```
kubectl label node $NODE-NAME feature.node.kubernetes.io/network-sriov.capable=true
```

> That will deploy a sriov-network-config daemon in such nodes that will collect information about the network interfaces. That information is available through the `sriovnetworknodestates` crd. A couple of minutes after the deployment, there will be one sriovnetworknodestates per node with the name of the nodes.

Reviewer: ```suggestion
That will deploy a sriov-network-config daemonset pod to labeled nodes to collect information about the network interfaces. That information is available through the `sriovnetworknodestates` Custom Resource Definition. A couple of minutes after the deployment, there will be one `sriovnetworknodestates` resource per node, with the name of the node as the resource name.
```

> For more information about how to use the sriov operator, please refer to [sriov-network-operator](https://github.com/k8snetworkplumbingwg/sriov-network-operator/blob/master/doc/quickstart.md#configuration)

Reviewer: ```suggestion
For more information about how to use the SR-IOV operator, please refer to [sriov-network-operator](https://github.com/k8snetworkplumbingwg/sriov-network-operator/blob/master/doc/quickstart.md#configuration)
```

# Using Multus

Warning: this is for advanced users

[Multus CNI](https://github.com/k8snetworkplumbingwg/multus-cni) is a plugin that enables attaching multiple network interfaces to pods. Multus does not replace CNI plugins, instead it acts as a CNI plugin multiplexer. Multus is useful in certain use cases, especially when pods are network intensive and require extra network interfaces that support dataplane acceleration techniques such as SRIOV.

Multus should be deployed after the cluster is already up and running and has a CNI plugin taking care of the cluster networking, e.g. Canal. After installing Multus, the current CNI plugin becomes the main CNI plugin under Multus transparently and thus, the cluster can be operated as before, as all pods will get an interface from the main plugin automatically. For more information refer to the [multus-cni](https://github.com/k8snetworkplumbingwg/multus-cni/tree/master/docs)

To deploy Multus, use following config:

```
apiVersion: helm.cattle.io/v1
kind: HelmChart
metadata:
  name: rke2-multus
  namespace: kube-system
spec:

>   repo: https://rke2-charts.rancher.io

Reviewer: I don't think we want to instruct users to add and pull from this repo directly - its currently just used as a staging point for the RKE2 build pipeline and we need to be careful if we're going to change that.

I think we should package the rke2-multus chart and have users enable it with a CLI flag as discussed at https://github.com/rancher/rke2/issues/746#issuecomment-832541123

Author: If user wants multus with default parameters, then he/she adds multus in the cni flag. 
If user wants multus without default parameters, how can he/she change them without pulling from the repo? Or we don't allow this through rke2 and it needs to happen via Rancher?

  chart: rke2-multus
```

The config needs to be copied over to the manifests directory:

```
mkdir -p /var/lib/rancher/rke2/server/manifests/
cp rke2-multus.yml /var/lib/rancher/rke2/server/manifests/
```

For more information about the options of the Multus config please refer to the [rke2-charts](https://github.com/rancher/rke2-charts/blob/main-source/packages/rke2-multus/charts/values.yaml)

## Using Multus with the containernetworking plugins

To take advantage of Multus, extra CNI plugins should be used to manage the additional network interfaces attached per pod. That is why, when installing the Multus helm chart, the common CNI plugins maintained by the containernetworking team are deployed (e.g. macvlan CNI). For more information about these plugins, check the [containernetworking plugins](https://www.cni.dev/plugins/current).

> As explained in the Multus documentation, to use one of these plugins, a NetworkAttachmentDefinition object would need to be created first pointing to one of these CNI plugins. After that, in the pod annotations, we can refer to it so that Multus can start the process to provide one extra interface to that pod. Please follow this example that configures macvlan cni plugin with multus [example](https://github.com/k8snetworkplumbingwg/multus-cni/blob/master/docs/quickstart.md#storing-a-configuration-as-a-custom-resource)

Reviewer 2: `...follow this example...` -> `...use example as reference...`?

## Using Multus with SRIOV

Using the SRIOV-cni as part of Multus, can help with dataplane acceleration use cases, providing an extra interface in the pod that can achieve very high throughput. SRIOV will not work in all environments and there are several requirements that must be fulfilled to consider the node as sriov capable:

* Physical NIC must support SRIOV (e.g. by checking /sys/class/net/$NIC/device/sriov_totalvfs)
* The host operating system must activate IOMMU virtualization

> * The vfio driver must be available in the host operating system

Reviewer 2: IOMMU & vfio would only be needed for DPDK, right? Should we clarify?

Author: I think IOMMU is needed for SRIOV. vfio is not, you are right, there might be alternatives

> SRIOV requires two different Helm charts. One for the crds:

Reviewer 2: `...in addition to Multus...`

```
apiVersion: helm.cattle.io/v1
kind: HelmChart
metadata:
  name: rke2-sriov-crd
  namespace: kube-system
spec:
  repo: https://rke2-charts.rancher.io
  chart: rke2-sriov-crd
```

And one for the sriov operator:

```
apiVersion: helm.cattle.io/v1
kind: HelmChart
metadata:
  name: rke2-sriov

>   namespace: kube-system

Reviewer 2: Should we recommend deploying this on it's own namespace?

Author: I think it is fine to deploy HelmCharts in kube-system. That does not mean the deployment will end up there too

spec:
  repo: https://rke2-charts.rancher.io
  chart: rke2-sriov
```

The config needs to be copied over to the manifests directory:

```
cp rke2-sriov-crd.yml rke2-sriov.yml /var/lib/rancher/rke2/server/manifests/
```

For more information about the options of the sriov config please refer to the [rke2-charts](https://github.com/rancher/rke2-charts/blob/main-source/packages/rke2-sriov/charts/values.yaml)

After installing the sriov charts, the sriov operator will be deployed. Then, the user must specify what nodes in the cluster are sriov capable by providing the label feature.node.kubernetes.io/network-sriov.capable=true:

```
kubectl label node $NODE-NAME feature.node.kubernetes.io/network-sriov.capable=true
```

That will deploy a sriov-network-config daemon in such nodes that will collect information about the network interfaces. That information is available thorugh the `sriovnetworknodestates` crd. A couple of minutes after the deployment, there will be one sriovnetworknodestates per node with the name of the nodes.

> 

Reviewer 2: Should we comment on webhooks and certificates?
